include/ringbuffer_utils.hpp

```diff
@@ -152,6 +153,151 @@ class PerfRingBufferReader {
   uint64_t _head;
 };
 
+struct MPSCRingBufferHeader {
+  uint64_t size;
+  static constexpr uint64_t k_discard_bit = 1UL << 62;
+  static constexpr uint64_t k_busy_bit = 1UL << 63;
+
+  static bool is_busy(uint64_t size) { return size & k_busy_bit; }
+  static bool is_discarded(uint64_t size) { return size & k_discard_bit; }
+};
+
+class MPSCRingBufferWriter {
```

Comment:
I think I understand.  In the common setting, the underlying RingBuffer is going to be shared--either in a `mmap(...MAP_SHARED);` or in a separate thread (multiprocess is the worst-case, so I'll focus on that).  On the other hand, the MPSCRingBufferWriter/Reader classes are _views_ into the Ringbuffer, with all private variables intended to be local to the current thread.

Reply:
The object model is not clear because there is not object that represents the ring buffer and is responsible for its creation/destruction.
`RingBuffer` holds the data necessary to access the ring buffer but does not act on it. The only requirement is a zero-initialized chunk of memory. And it is indeed shared explicitly (several threads accessing the same object) or implicitly (pointers in different `RingBuffer` objects pointing to the same underlying memory).
Writer/Reader are not really views (IMHO `RingBuffer` is the view) but local short-lived objects that modify the state of the ring buffer in a consistent manner.